File: Python/Decorator/Decorator.py
#! /usr/bin/python
# -*- coding: utf-8 -*-
'''
Decorator Pattern
Author: Kei Nakata
Data: Oct.10.2014
'''
import abc
import exceptions

class Component(object):

    def operation(self):
        print "I want to eat curry"

class ConcreteComponent(Component):

    def operation(self):
        print "I want to eat curry rice"

class Decorator(object):
    __metaclass__ = abc.ABCMeta
    def __init__(self):
        self.component = Component()

    @abc.abstractmethod
    def operation(self):
        pass

class ConcreteDecorator(Decorator):

    def operation(self):
        self.component.operation()
        print "and nann!!"

if __name__ == '__main__':
    a = ConcreteComponent()
    b = ConcreteDecorator()
    print "A:"
    a.operation()
    print "B:"
    b.operation()

File: C++/Decorator/Decorator.cpp
/*
 * Decorator Pattern
 * Author: Kei Nakata
 * Data: Oct.6.2014
 * Attach additional responsibilities to an object dynamically.
 * Decorators provide a flexible alternative to subclassing for
 * extending functionality.
 */
#include <iostream>
#include <memory>
#include <cstring>
#include <cstdlib>
#include <stack>
using std::cout;
using std::endl;
using std::string;
using std::stack;
using std::shared_ptr;
using std::move;

class Component
{
public:
    virtual ~Component() = default;
    virtual void operation() = 0;
};

class ConcreteComponet : public Component
{
public:
    void operation() override {
        cout << "I want to eat curry rice" << endl;
    }
};

class Decorator : public Component
{
protected:
    shared_ptr<Component> component_;
public:
    Decorator(shared_ptr<Component>& component): component_ {component} {}
    virtual ~Decorator() override = default;
    virtual void operation() = 0;
};

class ConcreteDecorator : public Decorator
{
public:
    using Decorator::Decorator;
    virtual ~ConcreteDecorator() override = default;
    void operation() override {
        component_ -> operation();
        cout << "and nann!!" << endl;
    }
};

int main()
{
    shared_ptr<Component> a {new ConcreteComponet};
    ConcreteDecorator b {a};
    cout << "A:" << endl;
    a->operation();
    cout << "B:" << endl;
    b.operation();

    return EXIT_SUCCESS;
}
